src/onnx/parse_scan.cpp

```diff
@@ -0,0 +1,314 @@
+    void modify_body(module_ref mod,
+                     const std::vector<instruction_ref>& args,
+                     int64_t n,
+                     int64_t m,
+                     const std::vector<int64_t>& scan_input_axes,
+                     const std::vector<int64_t>& scan_input_directions) const
+    {
+        std::vector<instruction_ref> params;
+        params.reserve(n + m);
+        transform(mod->get_parameter_names(),
+                  std::back_inserter(params),
+                  [&](const std::string& name) { return mod->get_parameter(name); });
+
+        // iteration_num, condition, and duplicate loop_state_variables are appended to parameters.
+        // References to the original loop_state_variables in other instructions are then replaced
+        // with references to the duplicate ones, after which the originals are removed.
+        //
+        // References to the scan_input_slices are replaced with references to inserted
+        // scan_slice->squeeze instructions, after which the scan_input_slices parameters are
+        // removed.
+        auto iter_param = mod->add_parameter("iter", shape{shape::int64_type});
+        auto cond_param = mod->add_parameter("cond", shape{shape::bool_type});
+        std::vector<instruction_ref> new_params;
+        new_params.reserve(n);
+        for(auto i = 0; i < n; ++i)
+            new_params.push_back(
+                mod->add_parameter("state_var" + std::to_string(i), params[i]->get_shape()));
+
+        for(auto i = 0; i < params.size(); ++i)
+        {
+            for(auto ins : iterator_for(*mod))
+            {
+                if(not contains(ins->inputs(), params[i]))
+                    continue;
+
+                auto new_ins = i < n ? new_params[i] : args[i];
+                if(i >= n)
+                {
+                    auto scan_axis = scan_input_axes[i - n];
+                    auto scan_dir  = scan_input_directions[i - n];
+                    new_ins        = mod->insert_instruction(
+                        params[i],
+                        make_op("scan_slice", {{"axis", scan_axis}, {"direction", scan_dir}}),
+                        new_ins,
+                        iter_param);
+                    new_ins = mod->insert_instruction(
+                        params[i], make_op("squeeze", {{"axes", {scan_axis}}}), new_ins);
+                }
+                ins->replace_argument(ins, params[i], new_ins);
+            }
+            mod->remove_instruction(params[i]);
+        }
+
+        auto returns = mod->get_returns();
+        returns.insert(returns.begin(), cond_param);
+        mod->replace_return(returns);
+    }
+
+    std::vector<int64_t> make_perm_for_scan_out(int64_t rank, int64_t axis) const
+    {
+        std::vector<int64_t> perm(rank);
+        std::iota(perm.begin(), perm.end(), 0);
+        std::copy(perm.begin() + 1, perm.begin() + 1 + axis, perm.begin());
```

Comment:
I wonder if `begin() + 1 + axis` would go out of bound if axis is the last axis.

Reply:
For rank r, the last axis would be r-1, meaning the range would be [begin + 1, begin + r). Since the perm vector has r elements, begin + r is the first element beyond the last element, making it the valid end iterator.